File: experimentos/piloto/sr/CambioClave.cs
using Solucion;
using System;

namespace Operaciones;

public class CambioClave
{
    private static RespuestaCaminoFeliz CambioClaveOk(string dni, string nueva_clave, Estado estado)
    {
        foreach (var clave in estado.Claves)
        {
            if (clave.Dni == dni)
            {
                clave.Clave = nueva_clave;
            }
        }

        estado.Movimientos.Add(new MovimientoModel()
        {
            Dni = dni,
            Movimiento = new Movimiento()
            {
                Fecha = Constantes.AHORA,
                Operacion = OperacionTipo.CLAVE,
            }
        });


        return new RespuestaCaminoFeliz()
        {
            NuevoEstado = estado,
            ValorParaMensaje = nueva_clave,
        };
    }
    public static RespuestaOperacion Ejecutar(string dni, string clave, string nueva_clave, Estado estado)
    {
        var respuestaOperacion = new RespuestaOperacion()
        {
            ResultadoOperacion = true,
            MensajeRespuesta = "",
            NuevoEstado = estado,
        };

        var respuestaError = Errores.UsuarioInexistente(dni, estado.Usuarios);
        if (respuestaError.ContieneError)
        {
            respuestaOperacion.ResultadoOperacion = false;
            respuestaOperacion.MensajeRespuesta = respuestaError.MensajeError;
            Console.WriteLine(respuestaOperacion.MensajeRespuesta);
            return respuestaOperacion;
        }

        respuestaError = Errores.ClaveIncorrecta(dni, clave, estado.Claves);
        if (respuestaError.ContieneError)
        {
            respuestaOperacion.ResultadoOperacion = false;
            respuestaOperacion.MensajeRespuesta = respuestaError.MensajeError;
            Console.WriteLine(respuestaOperacion.MensajeRespuesta);
            return respuestaOperacion;
        }

        respuestaError = Errores.CambioDeClaveBloqueado(dni, estado.Movimientos);
        if (respuestaError.ContieneError)
        {
            respuestaOperacion.ResultadoOperacion = false;
            respuestaOperacion.MensajeRespuesta = respuestaError.MensajeError;
            Console.WriteLine(respuestaOperacion.MensajeRespuesta);
            return respuestaOperacion;
        }

        respuestaError = Errores.NoCumpleRequisitosClave1(nueva_clave);
        if (respuestaError.ContieneError)
        {
            respuestaOperacion.ResultadoOperacion = false;
            respuestaOperacion.MensajeRespuesta = respuestaError.MensajeError;
            Console.WriteLine(respuestaOperacion.MensajeRespuesta);
            return respuestaOperacion;
        }

        respuestaError = Errores.NoCumpleRequisitosClave2(nueva_clave);
        if (respuestaError.ContieneError)
        {
            respuestaOperacion.ResultadoOperacion = false;
            respuestaOperacion.MensajeRespuesta = respuestaError.MensajeError;
            Console.WriteLine(respuestaOperacion.MensajeRespuesta);
            return respuestaOperacion;
        }

        var respuestaCaminoFeliz = CambioClaveOk(dni, nueva_clave, estado);
        respuestaOperacion.MensajeRespuesta = Resultado.OK + "Nueva clave: " + respuestaCaminoFeliz.ValorParaMensaje;
        respuestaOperacion.NuevoEstado = respuestaCaminoFeliz.NuevoEstado;

        // En respuestaOperacion.NuevoEstado tenemos nueestro estado actualizado. A fines prácticos mostramos solo el mensaje de respuesta de la operación.
        Console.WriteLine(respuestaOperacion.MensajeRespuesta);
        return respuestaOperacion;
    }
}

File: solucion_modelo/operaciones/clave/cambio_clave.py
from modelos.movimiento import Movimiento
from operaciones.clave.parametros_cambio_clave import ParametrosCambioClave
from tipos import RESULTADO, OPERACION
from estado import Estado
from constantes import CANTIDAD_CAMBIOS_CLAVE_MES, ahora
from utiles import (
    cantidad_cambios_clave_usuario_mes,
    contiene_letras_numeros,
)


def cambio_clave(solicitud: ParametrosCambioClave):

    estado = Estado()

    if solicitud.dni not in estado.usuarios.keys():
        return RESULTADO.UsuarioInexistente

    if estado.usuarios[solicitud.dni].clave != solicitud.clave:
        return RESULTADO.ClaveIncorrecta

    if len(solicitud.nueva_clave) < 8:
        return RESULTADO.NoCumpleRequisitosClave1

    if not contiene_letras_numeros(solicitud.nueva_clave):
        return RESULTADO.NoCumpleRequisitosClave2

    if (
        cantidad_cambios_clave_usuario_mes(solicitud.dni, estado.movimientos)
        > CANTIDAD_CAMBIOS_CLAVE_MES
    ):
        return RESULTADO.CambioDeClaveBloqueado

    estado.usuarios[solicitud.dni].clave = solicitud.nueva_clave

    estado.movimientos.append(Movimiento(ahora, OPERACION.CLAVE, solicitud.dni))

    estado.guardar()

    return RESULTADO.OK
